app/org/sagebionetworks/bridge/services/ScheduledActivityService.java

```diff
@@ -180,13 +180,68 @@ public ScheduledActivityList getActivityHistory(String healthCode,
     }
 
     public List<ScheduledActivity> getScheduledActivities(ScheduleContext context) {
-        checkNotNull(context);
-        return getScheduledActivities(context, V3_FILTER, V3_MERGE);
+        List<ScheduledActivity> scheduledActivities = scheduleActivitiesForPlans(context);
+        List<ScheduledActivity> dbActivities = activityDao.getActivities(context.getEndsOn().getZone(), scheduledActivities);
+        
+        Map<String, ScheduledActivity> dbMap = Maps.uniqueIndex(dbActivities, ScheduledActivity::getGuid);
+        
+        List<ScheduledActivity> saves = performMerge(scheduledActivities, dbMap, V3_MERGE);
+        activityDao.saveActivities(saves);
+        
+        return orderActivities(scheduledActivities, V3_FILTER);
     }
     
     public List<ScheduledActivity> getScheduledActivitiesV4(ScheduleContext context) {
-        checkNotNull(context);
-        return getScheduledActivities(context, V4_FILTER, V4_MERGE);
+        List<ScheduledActivity> scheduledActivities = scheduleActivitiesForPlans(context);
+
+        // Get all persisted activities within the time frame, not just those found by the scheduler (as in v3).
+        Map<String, ScheduledActivity> dbMap = retrieveAllPersistedActivitiesIntoMap(context, scheduledActivities);
+        
+        // Compare scheduled and persisted activities, replacing scheduled with persisted where they exist
+        List<ScheduledActivity> saves = performMerge(scheduledActivities, dbMap, V4_COLLECT_SAVES);
+        activityDao.saveActivities(saves);
+        
+        // We've removed all the persisted activities that were found by the scheduler. Those have been saved
+        // as necessary. The remaining tasks were scheduled based on user-triggered events that can be updated, 
+        // not events like "enrollment" but events like "when this activity is finished." These are now all 
+        // added to the activities that will be returned.
+        scheduledActivities.addAll(dbMap.values());
```

Comment:
Expired tasks can run into the hundreds or possibly thousands in the V3 API, if the study is long-running. I don't think we should persist them in this API.

I can add logic to find *persisted* activities that are safe to update, and recheck them to see if they are up-to-date, and then save those, before adding them into the merge.